src/misclibs.F

```diff
@@ -3630,6 +3644,210 @@ subroutine calcksmax(nstat,rstat,dt,uh,vh,mf,kmh,kmv,khh,khv)
+
+      dtdx=dt*rdx*rdx
+      dtdy=dt*rdy*rdy
+      dtdz=dt*rdz*rdz
+
+!$acc parallel loop gang vector collapse(2) default(present)
+      do j=1,nj
+      do i=1,ni
+          index=(j-1)*ni+i
+          ksh(index) = -99999.0
+          ksv(index) = -99999.0
+      enddo
+      enddo
+
+!$omp parallel do default(shared)  &
+!$omp private(i,j,k,tem)
+
+!$acc parallel default(present) 
+!$acc loop vector collapse(2) 
+      do j=1,nj
+      do i=1,ni
+!$acc loop seq private(index)
+         do k=2,nk
+            index=(j-1)*ni+i
+            tem = khh(i,j,k)*dtdx*uh(i)*uh(i)
+            ksh(index)=max(tem,ksh(index))
+
+            tem = khh(i,j,k)*dtdy*vh(j)*vh(j)
+            ksh(index)=max(tem,ksh(index))
+
+            tem = khv(i,j,k)*dtdz*mf(i,j,k)*mf(i,j,k)
+            ksv(index)=max(tem,ksv(index))
+         end do
+      end do
+      enddo
+!$acc end parallel
+
+!$omp parallel do default(shared)  &
+!$omp private(i,j,k,tem)
+!$acc parallel loop gang vector collapse(2) default(present)
+      do i=1,ni
+         do j=1,nj
+!$acc loop seq private(index)
+            do k=2,nk
+               index=(j-1)*ni+i
+               tem = khh(i,j,k)*dtdx*uh(i)*uh(i)
+               if( tem.eq.ksh(index) )then
+                 imaxth(index)=i
+                 jmaxth(index)=j
+                 kmaxth(index)=k
+               end if
+               tem = khh(i,j,k)*dtdy*vh(j)*vh(j)
+               if( tem.eq.ksh(index) )then
+                 imaxth(index)=i
+                 jmaxth(index)=j
+                 kmaxth(index)=k
+               end if
+               tem = khv(i,j,k)*dtdz*mf(i,j,k)*mf(i,j,k)
+               if( tem.eq.ksv(index) )then
+                 imaxtv(index)=i
+                 jmaxtv(index)=j
+                 kmaxtv(index)=k
+               endif
+            end do
+         end do
+      end do
+
+      fhmax=-99999999.
+      fvmax=-99999999.
+      imaxh=1
+      jmaxh=1
+      kmaxh=1
+      imaxv=1
+      jmaxv=1
+      kmaxv=1
+
+!! Using a serial loop here so all the data is sync'd between consecutive kernel
+!calls.
+!! The RANGE = PAIRS or (PAIRS+1) as a way to manage odd numbers.
+      PAIRS=ni*nj/2           !! Round down.
+      RANGE=(ni*nj+1)/2       !! Round up.
+      do while ( RANGE > 1 )
+!! Now fold the 2-d array with half the threads each time.
+!$acc parallel loop gang vector private(i)
+        do i=1,PAIRS
+          if( ksh(i) .le. ksh(RANGE)) then
+             ksh(i) = ksh(RANGE+1)
+             imaxth(i) = imaxth(RANGE+1)
+             jmaxth(i) = jmaxth(RANGE+1)
+             kmaxth(i) = kmaxth(RANGE+1)
+          endif
+          if( ksv(i) .le. ksv(RANGE)) then
+             ksv(i) = ksv(RANGE+1)
+             imaxtv(i) = imaxtv(RANGE+1)
+             jmaxtv(i) = jmaxtv(RANGE+1)
+             kmaxtv(i) = kmaxtv(RANGE+1)
+          endif
+        enddo
+        PAIRS=RANGE/2
+        RANGE=(RANGE+1)/2
+      enddo 
+
```

Comment:
I think an `acc update host (ksh,imaxth,jmaxth,kmaxth,ksv,imaxtv,jmaxtv,kmaxtv)` directive is needed here.

Reply:
Thanks @sjsprecious for catching this.  I believe you're correct and I'll add the update host and verify correctness.

Reply:
calccfl_fold will need this change as well

Reply:
Both of these updates to the host were added in [e00ae25](https://github.com/george-bryan/CM1/pull/53/commits/e00ae254d365a73a0f2eaee2381c989be5fd04ff)